tests/test_ims.py

```diff
@@ -1,38 +1,111 @@
-import imp
 import unittest
 
-import os
+import click
+import re
 
 import ims.common.config as config
 
 config.load()
 
 
-def find_test_modules(skip_list):
-    dir_name, cur_name = os.path.split(__file__)
-    name_of_this_module, cur_ext = os.path.splitext(cur_name)
-    modules = []
-    for dirname, dirnames, filenames in os.walk(dir_name):
-        for filename in filenames:
-            name, ext = os.path.splitext(filename)
-            if ext == '.py' and name != '__init__' \
-                    and name != name_of_this_module \
-                    and name not in skip_list:
-                path = os.path.join(dirname, filename)
-                modules.append(imp.load_source(name, path))
-    return modules
+def get_test_cases(suite):
+    """
+    Returns a list of Test Cases that are part of the given Test Suite
 
+    :param suite: The Test Suite whose Test Cases needs to be extracted
+    :return: A List of Test Cases
+    """
+    test_cases = set()
+    for test in suite:
+        if not isinstance(test, unittest.TestSuite):
+            test_cases.add(test)
+        else:
+            results = get_test_cases(test)
+            test_cases = test_cases.union(results)
+    return test_cases
 
-skip_list = []
 
-modules = find_test_modules(skip_list)
+def fetch_all_tests():
+    """
+    Loads and returns all test cases in the tests directory
+
+    :return: A List of Test Cases
+    """
+    loader = unittest.TestLoader()
+    tests = loader.discover('tests')
+    return get_test_cases(tests)
+
+
+def get_class_name(cls):
+    """
+    Returns the fully qualified name of the class
+
+    :param cls: the class whose name needs to be returned
+    :return: Class Name as String
+    """
+    return str(cls)[8:-2]
+
+
+def filter_tests(pattern, tests):
+    """
+    Returns a list of test cases that match the given pattern
+
+    :param pattern: The regex pattern that needs to be matched
+    :param tests: The list of test cases that needs to be filtered
+    :return: A List of test cases
+    """
+    filtered_tests = set()
+    for test in tests:
+        if re.search(pattern, get_class_name(test.__class__)) is not None:
+            filtered_tests.add(test)
+    return filtered_tests
+
+
+def run_tests(tests):
+    """
+    Runs the given tests
+
+    :param tests: The List of Test Cases to run
+    :return: None
+    """
+    suite = unittest.TestSuite(tests)
+    test_runner = unittest.TextTestRunner(verbosity=3)
+    test_runner.run(suite)
+
+
+@click.group()
+def cli():
+    """
+    The Testing Framework for BMI, do test_ims run --help for more details
+    """
+
+
+@cli.command(short_help="Runs the test cases as per the given patterns")
+@click.argument('test_case_patterns', nargs=-1)
+@click.option('-i', '--interactive', is_flag=True, help="Interactive Mode")
+def run(test_case_patterns, interactive):
```

Comment:
@chemistry-sourabh setup is called for every test `case` whereas __init__ is called for every Test `Class`.

Like I said, I don't have a strong opinion, do whichever. Just that since there are unittest test cases in the code, we should probably use the unittest test runner instead of the pytest test runner because pytest may not always support all of unittest frameworks test features. We decide one test framework and stick to it.